extensions/amp-a4a/0.1/amp-a4a.js

```diff
@@ -680,6 +702,50 @@ export class AmpA4A extends AMP.BaseElement {
   }
 
   /**
+   * Measures the amount of time until last ad url generation for all amp-ad
+   * elements of this type.
+   * @private
+   */
+  measureGetAdUrlsDelay_() {
+    const startTime = Date.now();
+    const type = this.element.getAttribute('type');
+    const promiseId = this.promiseId_;
+    // Only execute once per page, this includes potential pauseCallback flows.
+    if (this.win['a4a-measuring-ad-urls']) {
+      return;
+    }
+    this.win['a4a-measuring-ad-urls'] =
+     (() => resourcesForDoc(this.element).getMeasuredResources(this.win,
+       r => {
+         return r.element.tagName == 'AMP-AD' &&
+           r.element.getAttribute('type') == type;
+       })
+     .then(resources => {
+       if (promiseId != this.promiseId_) {
+         throw cancellation();
+       }
+       const getAdUrlsPromise = [];
+       resources.forEach(r => getAdUrlsPromise.push(r.element.getImpl().then(
+         instance => {
+           //
```

Comment:
Updated to comment:  
// Note that ad url promise could be null if isValidElement returns false.